entrypoint.go

```diff
@@ -0,0 +1,426 @@
+	// Wait blocks until the specified file exists or the context is done.
+	Wait(ctx context.Context, file string, expectContent bool, breakpointOnFailure bool) error
+}
+
+// Runner encapsulates running commands.
+type Runner interface {
+	Run(ctx context.Context, args ...string) error
+}
+
+// PostWriter encapsulates writing a file when complete.
+type PostWriter interface {
+	// Write writes to the path when complete.
+	Write(file, content string)
+}
+
+// Go optionally waits for a file, runs the command, and writes a
+// post file.
+func (e Entrypointer) Go() error {
+	prod, _ := zap.NewProduction()
+	logger := prod.Sugar()
+
+	_, err := writeToTempFile(postFile)
+	if err != nil {
+		return err
+	}
+
+
+	output := []result.RunResult{}
+	defer func() {
+		if wErr := termination.WriteMessage(e.TerminationPath, output); wErr != nil {
+			logger.Fatalf("Error while writing message: %s", wErr)
+		}
+		_ = logger.Sync()
+	}()
+
+	if err := os.MkdirAll(filepath.Join(e.StepMetadataDir, "results"), os.ModePerm); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(filepath.Join(e.StepMetadataDir, "artifacts"), os.ModePerm); err != nil {
+		return err
+	}
+	
+	for _, f := range e.WaitFiles {
+		if err := e.Waiter.Wait(context.Background(), f, e.WaitFileContent, e.BreakpointOnFailure); err != nil {
+			// An error happened while waiting, so we bail
+			// *but* we write postfile to make next steps bail too.
+			// In case of breakpoint on failure do not write post file.
+			if !e.BreakpointOnFailure {
+				e.WritePostFile(e.PostFile, err)
+			}
+			output = append(output, result.RunResult{
+				Key:        "StartedAt",
+				Value:      time.Now().Format(timeFormat),
+				ResultType: result.InternalTektonResultType,
+			})
+
+			if errors.Is(err, ErrSkipPreviousStepFailed) {
+				output = append(output, e.outputRunResult(pod.TerminationReasonSkipped))
+			}
+
+			return err
+		}
+	}
+
+	output = append(output, result.RunResult{
+		Key:        "StartedAt",
+		Value:      time.Now().Format(timeFormat),
+		ResultType: result.InternalTektonResultType,
+	})
+
+	var err error
+	if e.Timeout != nil && *e.Timeout < time.Duration(0) {
+		err = errors.New("negative timeout specified")
+	}
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	if err == nil {
+		if err := e.applyStepArtifactSubstitutions(pipeline.StepsDir); err != nil {
+			logger.Error("Error while substituting step artifacts: ", err)
+		}
+
+		ctx, cancel = context.WithCancel(ctx)
+		if e.Timeout != nil && *e.Timeout > time.Duration(0) {
+			ctx, cancel = context.WithTimeout(ctx, *e.Timeout)
+		}
+		defer cancel()
+		// start a goroutine to listen for cancellation file
+		go func() {
+			if err := e.waitingCancellation(ctx, cancel); err != nil && (!IsContextCanceledError(err) && !IsContextDeadlineError(err)) {
+				logger.Error("Error while waiting for cancellation", zap.Error(err))
+			}
+		}()
+		err = e.Runner.Run(ctx, e.Command...)
+	}
+
+	var ee *exec.ExitError
+	switch {
```

Comment:
The error handling in the `Go` function could be improved. Currently, it's a bit scattered and inconsistent. Consider creating a separate function for error handling to make the code cleaner and easier to maintain. [medium]